Reflect derive is broken on main for empty enums

## Bevy version

Commit: ec822c8c

## What you did
```rs
#[derive(Reflect)]
enum Empty {}
```

## What went wrong
`cargo build`:
```
error[E0277]: `!` does not implement `Reflect` so cannot be fully reflected
  --> examples/reflection/reflection.rs:51:10
   |
51 | #[derive(Reflect)]
   |          ^^^^^^^ the trait `bevy_reflect::Reflect` is not implemented for `!`
   |
   = note: consider annotating `!` with `#[derive(Reflect)]`
   = help: the following other types implement trait `bevy_reflect::Reflect`:
             &'static std::panic::Location<'static>
             &'static std::path::Path
             &'static str
             ()
             (A, B)
             (A, B, C)
             (A, B, C, D)
             (A, B, C, D, E)
           and 735 others
   = note: this error might have been caused by changes to Rust's type-inference algorithm (see issue #48950 <https://github.com/rust-lang/rust/issues/48950> for more information)
   = help: did you intend to use the type `()` here instead?
   = note: required for the cast from `Box<!>` to `Box<dyn bevy_reflect::Reflect>`
   = note: this error originates in the derive macro `Reflect` (in Nightly builds, run with -Z macro-backtrace for more info)
```
rust-analayzer:
```
missing match arm: type `&Empty` is non-empty
```

## Additional information

Seem that the problem is in this code generated by `Reflect` derive macro:
```rs
#[inline]
#[allow(
    unreachable_code,
    reason = "Ignored fields without a `clone` attribute will early-return with an error"
)]
fn reflect_clone(
    &self,
) -> ::core::result::Result<
    bevy_reflect::__macro_exports::alloc_utils::Box<dyn bevy_reflect::Reflect>,
    bevy_reflect::ReflectCloneError,
> {
    let this = self;
    ::core::result::Result::Ok(bevy_reflect::__macro_exports::alloc_utils::Box::new(
        match this {},
    ))
}
```

--- a/crates/bevy_reflect/derive/src/derive_data.rs
+++ b/crates/bevy_reflect/derive/src/derive_data.rs
@@ -937,7 +937,10 @@ impl<'a> ReflectEnum<'a> {
             }
         };
 
-        let body = if self.meta.is_remote_wrapper() {
+        let body = if variant_patterns.is_empty() {
+            // enum variant is empty, so &self will never exist
+            quote!(unreachable!())
+        } else if self.meta.is_remote_wrapper() {
             quote! {
                 let #this = <Self as #bevy_reflect_path::ReflectRemote>::as_remote(self);
                 #FQResult::Ok(#bevy_reflect_path::__macro_exports::alloc_utils::Box::new(<Self as #bevy_reflect_path::ReflectRemote>::into_wrapper(#inner)))

--- a/crates/bevy_reflect/src/lib.rs
+++ b/crates/bevy_reflect/src/lib.rs
@@ -2644,6 +2644,14 @@ bevy_reflect::tests::Test {
     }
 
     #[test]
+    fn should_allow_empty_enums() {
+        #[derive(Reflect)]
+        enum Empty {}
+
+        assert_impl_all!(Empty: Reflect);
+    }
+
+    #[test]
     fn recursive_typed_storage_does_not_hang() {
         #[derive(Reflect)]
         struct Recurse<T>(T);

